### ISSUE TITLE
GcsSession.readRaw Does Not Play Nice With MessageSourceMutator afterReceive (for deleting) When Streaming

### ISSUE THREAD

**Issue by fiidim** [GoogleCloudPlatform/spring-cloud-gcp#1312]
Please bear with me on this one.

I'm creating a supplier `stream-application` for `spring-cloud-dataflow` that picks up files from a GCS bucket, and emits the data contained therein.  The supplier function is built almost **verbatim** from the existing `sftp stream-application`.  When the file is received, there is a `org.springframework.integration.aop.MessageSourceMutator` hook that will remove the file (`FileDeletingMessageAdvice`), then a subsequent step that reads the channel and emits the data. Most everything works correctly with `spring-integration` except for the following:

1. The repo is missing several files for full integration that I had to supplement (mostly dsl files like GcsInboundChannelAdapterSpec, GcsOutboundGatewaySpec, GcsStreamingInboundChannelAdapterSpec, and a class to return the adapters).  I can submit a PR/MR if you want.
2. When Streaming, the GcsSession.readRaw returns a ChannelnputStream with a BlobReadChannel.  When the `afterReceive` function of the `MessageSourceMutator` runs it removes the file.  Any subsequent attempts to read from the InputStream after this moment return a 404 error (since the file is no longer there).  Same thing happens during a rename.
```
2022-11-01 12:44:47.458 ERROR 65889 --- [oundedElastic-2] o.s.integration.handler.LoggingHandler   : org.springframework.messaging.MessageHandlingException: IOException while iterating; nested exception is java.io.IOException: com.google.cloud.RetryHelper$RetryHelperException: com.google.cloud.storage.StorageException: 404 Not Found
GET https://storage.googleapis.com/download/storage/v1/b/<bucket-name>/o/<filename>?alt=media
No such object: <bucket-name>/<filename>, failedMessage=GenericMessage [payload=sun.nio.ch.ChannelInputStream@1d6c033b, headers={file_remoteHostPort=storage.googleapis.com:443, file_remoteFileInfo={"directory":false,"filename":"<filename>","link":false,"modified":1667312728832,"permissions":"Use [BlobInfo.getAcl()] to obtain permissions.","remoteDirectory":"<bucket-name>","size":149619}, file_remoteDirectory=<bucket-name>, id=<id>, contentType=text/plain, closeableResource=org.springframework.integration.file.remote.session.CachingSessionFactory$CachedSession@6e67b353, file_remoteFile=<filename>, timestamp=1667321087373}]
	at org.springframework.integration.file.splitter.FileSplitter$FileIterator.hasNext(FileSplitter.java:344)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:133)
<snip>
Caused by: java.io.IOException: com.google.cloud.RetryHelper$RetryHelperException: com.google.cloud.storage.StorageException: 404 Not Found
GET https://storage.googleapis.com/download/storage/v1/b/<bucket-name>/o/<filename>?alt=media
No such object: <bucket-name>/<filename>
	at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:149)
	at java.base/sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65)
	at java.base/sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:107)
	at java.base/sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:101)
	at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:270)
	at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:313)
	at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:188)
	at java.base/java.io.InputStreamReader.read(InputStreamReader.java:177)
	at java.base/java.io.BufferedReader.fill(BufferedReader.java:162)
	at java.base/java.io.BufferedReader.readLine(BufferedReader.java:329)
	at java.base/java.io.BufferedReader.readLine(BufferedReader.java:396)
	at org.springframework.integration.file.splitter.FileSplitter$FileIterator.hasNextLine(FileSplitter.java:350)
	at org.springframework.integration.file.splitter.FileSplitter$FileIterator.hasNext(FileSplitter.java:334)
	... 74 more
```

`GcsSession` Code:
```
  @Override
  public InputStream readRaw(String source) throws IOException {
    String[] tokens = getBucketAndObjectFromPath(source);
    Assert.state(tokens.length == 2, "Can only write to files, not buckets.");
    return Channels.newInputStream(this.gcs.reader(tokens[0], tokens[1]));
  }
```

I presume that it's related to the nature of the `InputStream` being used, and when the data is made available to the stream.  It seems that with `sun.nio.ch.ChannelInputStream` the data might not be available until the first read, whereas with Apache's `SftpInputStreamAsync : InputStreamWithChannel` it is.

In summary, if the streaming `IntegrationFlow` where the file is deleted during `afterReceive` `org.springframework.integration.aop.MessageSourceMutator` is ever going to work, then a different `InputStream` or manipulation of the `InputStream` will need to occur.  If it's a coincidence that the `sftp stream-application` works, then that's bad.

**Comment by elefeint** [comment]
Just to confirm -- which parts of Spring Cloud GCP are you using in your poject -- is it the [Spring Integration adapters for GCS](https://googlecloudplatform.github.io/spring-cloud-gcp/3.4.0/reference/html/index.html#channel-adapters-for-google-cloud-storage)?

We never considered the Spring Cloud Dataflow usecase in this project, so not familiar with its fundamentals. It sounds like in your application the file gets deleted, and then there is an attempt to read its content? What project is the `MessageMutator` concept from -- is it Spring Cloud Dataflow?

**Comment by fiidim** [comment]
Yes, I leverage the `spring-cloud-gcp-storage` classes (mainly from `integration`), as well as `spring-integration-file`.

I don't think that SCDF is the important bit, it's more a problem with the `spring-integration` piece.  We are still building functional components (Supplier/Publisher) via Integration Flows.  The problem is that the standard flow for streaming a file has problems (create input stream, delete/rename, read from stream).  As a test, I read 1 byte from the input stream in order to poke it into life, and the flow finished correctly.

Sorry, I typo'd the Mutator.  It should have been `org.springframework.integration.aop.MessageSourceMutator' which is from `spring-integration-core`.  I will make the edit.

**Comment by fiidim** [comment]
Correct, however this behaviour is not consistent with other `org.springframework.integration.file.remote.session.Session` derived classes.  For example, `org.springframework.integration.ftp.session.FtpSession` which implements `Session<FTPFile>` does not exhibit this behaviour, nor does `org.springframework.integration.sftp.session.SftpSession` which implements `Session<SftpClient.DirEntry>`.  None of these need to have their first byte read in order to allow for reading after deletion.

I would assume that the majority of the *Session classes behave the correct way, and this one is anomalous.  

You can see examples [here](https://github.com/spring-cloud/stream-applications/tree/main/functions/supplier/ftp-supplier) and [here](https://github.com/spring-cloud/stream-applications/tree/main/functions/supplier/sftp-supplier).  These are purely supplier functions in a raw form that eventually get wrapped into an SCDF application.

**Comment by elefeint** [comment]
The difference is that ftp/sftp message source [first copies](https://github.com/spring-projects/spring-integration/blob/978724e212fbb3e4434216f7b79e4b5d36db287c/spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizingMessageSource.java#L267) the remote files to the local file system.
GCS adapters don't copy the file locally, instead streaming from the server directly. So if the storage object is deleted before reading begins, it understandably cannot be found.

**Comment by fiidim** [comment]
That's only true if you are using a `FileSynchronizing` adapter.   If you are using a streaming adapter, it does not local copy, and uses [AbstractRemoteFileStreamingMessageSource](https://github.com/spring-projects/spring-integration/blob/main/spring-integration-file/src/main/java/org/springframework/integration/file/remote/AbstractRemoteFileStreamingMessageSource.java).  See examples of `StreamingMessageSource` [here](https://github.com/spring-projects/spring-integration/blob/main/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/inbound/SftpStreamingMessageSource.java) and [here](https://github.com/spring-projects/spring-integration/blob/main/spring-integration-ftp/src/main/java/org/springframework/integration/ftp/inbound/FtpStreamingMessageSource.java).

The functions I reference before offer working in both modes.  Synchronizing or streaming.  In the context of this issue I reported, we are referring to streaming.

**Comment by elefeint** [comment]
@artembilan Could you help me understand whether Spring Integration deleting a remote file first, and then allowing the user to read the stream is supported behavior for streaming ftp/sftp?

**Comment by artembilan** [comment]
Well, we never advertised a remote file removal for a streaming message source: https://docs.spring.io/spring-integration/docs/current/reference/html/sftp.html#sftp-streaming.
Indeed there is a `InputStream` which might need to have a connection to the remote resource to give the data on demand.

I believe that it is just coincident that it works for FTP and SFTP as @fiidim states.

The code we talk about is here: https://github.com/spring-cloud/stream-applications/blob/main/functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpSupplierConfiguration.java#L223-L229
That advice is applied here: https://github.com/spring-cloud/stream-applications/blob/main/functions/supplier/sftp-supplier/src/main/java/org/springframework/cloud/fn/supplier/sftp/SftpSupplierConfiguration.java#L119-L137.

Which essentially does that `delete()` logic just after `messageSource.receive()`, but really before we process a returned message with an `InputStream` down the flow.

This is not a Spring Integration functionality by itself, therefore there is nothing to fix over there or probably even here for `GcsSession`. Unless you see some solution to keep the data in the other stream type before producing and therefore protecting ourselves from delete side-effect.

As a workaround for your `GcsSupplier` I'd suggest do not use that `FileDeletingMessageAdvice`, but rather take the respective info from message headers downstream and call `RemoteFileTemplate` manually.
Another workaround probably could be as an extra `MessageSourceMutator` impl which would produce for other another message instance with copied `inputStream` before `FileDeletingMessageAdvice` is called.
Or you can extend `RemoteFileDeletingAdvice` to do that logic before calling `super`: I believe you still declare the mentioned bean in your `GcsSupplierConfiguration`.

**Comment by fiidim** [comment]
@artembilan Thanks for your input on this.  I was hoping it was not a coincidence it worked with ftp/sftp, and that it may have been an implementation issue here.

There are definitely 2 approaches to getting the desired behaviour:
1. Duplicate the `InputStream` before the delete or rename happens.
2. Move where the delete/rename happens until after the `InputStream` is closed.  This seems like the correct approach.

For the first approach, I was able to keep the existing `MessagingAdvice`, and add a new one, ensuring it fires first.  The `afterReceive` method duplicates the `Stream`.
```
public class RemoteFileCopyStreamAdvice implements MessageSourceMutator {
	@Nullable
	@Override
	public Message<?> afterReceive(@Nullable Message<?> result, final MessageSource<?> source) {
		return result == null || !(result.getPayload() instanceof InputStream) ?
				result : convertStream(result);
	}

	public Message<?> convertStream(Message<?> result) {
		try {
			return MessageBuilder.withPayload(IOUtils.toBufferedInputStream((InputStream) result.getPayload()))
								.copyHeaders(result.getHeaders()).build();
		}
		catch (IOException e) {
		}
		return result;
	}
}
```

In the second approach, the `InputStream` is closed in the `FileSplitter`.  The delete or rename information would have to be provided to the `FileSplitter`, and the decision made in the close method.  I suspect it would be around [here](https://github.com/spring-projects/spring-integration/blob/7084e9654ac2d1d9c68a7fb3b6a4f0be426c0d68/spring-integration-file/src/main/java/org/springframework/integration/file/splitter/FileSplitter.java#L242)

**Comment by artembilan** [comment]
You know that code:
```
					Closeable closeableResource = StaticMessageHeaderAccessor.getCloseableResource(message);
					if (closeableResource != null) {
						closeableResource.close();
					}
```
Gives me a hint that `Closeable` we provide in a header could be not just an `InputStream`, but some wrapper which can perform a delete operation as well.
The `AbstractRemoteFileStreamingMessageSource` may be supplied with a `boolean deleteRemoteFile` which would lead to an extra logic.

Probably need to sleep with this idea.

Feel free to transfer this issue into https://github.com/spring-projects/spring-integration/issues since I believe there is nothing we can do in Spring Cloud GCP.

**Comment by elefeint** [comment]
@fiidim I am going to close this issue, please open a feature request in the [Spring Integration repo](https://github.com/spring-projects/spring-integration/issues).
